app/domain/authentication/authn_gcp/extract_resource_restrictions.rb

```diff
@@ -0,0 +1,88 @@
+require 'command_class'
+
+module Authentication
+  module AuthnGCP
+
+    # This class is responsible of restrictions extraction that are set on a Conjur host or user as annotations.
+    ExtractResourceRestrictions = CommandClass.new(
+      dependencies: {
+        role_class:     ::Role,
+        resource_class: ::Resource,
+        logger:         Rails.logger
+      },
+      inputs:       %i(account username extraction_prefix)
+    ) do
+
+      def call
+        extract_resource_restrictions
+        restrictions_list
+      end
+
+      private
+
+      def extract_resource_restrictions
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ExtractingRestrictionsFromResource.new(@username, @extraction_prefix))
+        prefixed_resource_annotations
+        init_restrictions_list
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ExtractedResourceRestrictions.new(restrictions_list.length()))
+      end
+
+      def restrictions_list
+        return @restrictions_list if @restrictions_list
+
+        @restrictions_list = Array.new
+      end
+
+      def prefixed_resource_annotations
+        @prefixed_resource_annotations ||= resource_annotations.select do |a|
+          annotation_name = a.values[:name]
+          annotation_name.start_with?(@extraction_prefix)
+        end
+      end
+
+      def init_restrictions_list
+        prefixed_resource_annotations.select do |a|
+          annotation_name = a.values[:name]
+          resource_value = annotation_value(annotation_name)
+          next unless resource_value
+          restrictions_list.push(
+            ResourceRestriction.new(
+              type: annotation_name,
+              value: resource_value
+            )
+          )
+        end
+      end
+
+      def resource_annotations
+        @resource_annotations ||= role.annotations
+
+        # This is an illegal state, because if the resource doesnt have annotations we should get an empty array
+        raise Errors::Conjur::FetchAnnotationsFailed.new(role_id) unless @resource_annotations
```

Comment:
this check will happen every time we call `resource_restrictions`. how about changing the memoization into 
```
return @resource_annotations if @resource_annotations

@resource_annotations ||= role.annotations
# This is an illegal state, because if the resource doesnt have annotations we should get an empty array
raise Errors::Conjur::FetchAnnotationsFailed.new(role_id) unless @resource_annotations
@resource_annotations
```